File: GRMCore/Class/cSetLandcover.vb
Public Class cSetLandcover

    Public Enum LandCoverCode
        WATR
        URBN
        BARE
        WTLD
        GRSS
        FRST
        AGRL
        CONST_VALUE
        NULL
    End Enum

    Public mLandCoverDataType As Nullable(Of cGRM.FileOrConst)
    Public mGridLandCoverFPN As String
    Public mLandCoverVATFPN As String
    Public mdtLandCoverInfo As GRMProject.LandCoverDataTable
    Public mConstRoughnessCoefficient As Nullable(Of Single)
    Public mConstImperviousRatio As Nullable(Of Single)

    Public Sub New()

    End Sub

    Public Sub SetValues(ByVal prjDB As GRMProject)
        If mLandCoverDataType.HasValue Then
            Dim row As GRMProject.ProjectSettingsRow = CType(prjDB.ProjectSettings.Rows(0), GRMProject.ProjectSettingsRow)
            With row
                .LandCoverDataType = mLandCoverDataType.Value.ToString
                Select Case mLandCoverDataType.Value
                    Case cGRM.FileOrConst.File
                        .LandCoverFile = mGridLandCoverFPN
                        .LandCoverVATFile = mLandCoverVATFPN
                        .SetConstantRoughnessCoeffNull()
                        .SetConstantImperviousRatioNull()
                    Case cGRM.FileOrConst.Constant
                        .SetLandCoverFileNull()
                        .SetLandCoverVATFileNull()
                        .ConstantRoughnessCoeff = CStr(mConstRoughnessCoefficient.Value)
                        .ConstantImperviousRatio = CStr(mConstImperviousRatio.Value)
                End Select
            End With
            If mLandCoverDataType.Equals(cGRM.FileOrConst.File) Then
                mdtLandCoverInfo.AcceptChanges()
                For Each r As DataRow In mdtLandCoverInfo.Rows
                    r.SetAdded()
                Next
            End If
        End If
    End Sub

    Public Sub GetValues(ByVal prjDB As GRMProject)
        mLandCoverDataType = Nothing
        mLandCoverVATFPN = Nothing
        mConstRoughnessCoefficient = Nothing
        mConstImperviousRatio = Nothing
        Dim row As GRMProject.ProjectSettingsRow = CType(prjDB.ProjectSettings.Rows(0), GRMProject.ProjectSettingsRow)
        With row
            If Not .IsLandCoverDataTypeNull Then
                Select Case .LandCoverDataType
                    Case cGRM.FileOrConst.File.ToString
                        mLandCoverDataType = cGRM.FileOrConst.File
                        mGridLandCoverFPN = .LandCoverFile
                        mLandCoverVATFPN = .LandCoverVATFile
                    Case cGRM.FileOrConst.Constant.ToString
                        mLandCoverDataType = cGRM.FileOrConst.Constant
                        mConstImperviousRatio = CSng(.ConstantImperviousRatio)
                        mConstRoughnessCoefficient = CSng(.ConstantRoughnessCoeff)
                    Case Else
                        Throw New InvalidOperationException
                End Select
            End If
        End With
        If mLandCoverDataType.Equals(cGRM.FileOrConst.File) Then
            mdtLandCoverInfo = New GRMProject.LandCoverDataTable
            mdtLandCoverInfo = prjDB.LandCover
        Else
            mdtLandCoverInfo = Nothing
        End If
    End Sub

    Public ReadOnly Property IsSet() As Boolean
        Get
            Return mLandCoverDataType.HasValue
        End Get
    End Property

    Public Function GetLandCoverName(ByVal intGridValue As Integer) As String
        If mLandCoverDataType.Equals(cGRM.FileOrConst.File) Then
            Dim rows As DataRow() = mdtLandCoverInfo.Select("GridValue = " & intGridValue)
            Return CStr(rows(0)("GRMLandCoverE"))
        Else
            Return "[CONST]"
        End If
    End Function

    Public Shared Function GetLandCoverCode(ByVal inName As String) As LandCoverCode
        Select Case Trim(inName)
            Case LandCoverCode.AGRL.ToString : Return cSetLandcover.LandCoverCode.AGRL
            Case LandCoverCode.BARE.ToString : Return cSetLandcover.LandCoverCode.BARE
            Case LandCoverCode.FRST.ToString : Return LandCoverCode.FRST
            Case LandCoverCode.GRSS.ToString : Return LandCoverCode.GRSS
            Case LandCoverCode.URBN.ToString : Return LandCoverCode.URBN
            Case LandCoverCode.WATR.ToString : Return LandCoverCode.WATR
            Case LandCoverCode.WTLD.ToString : Return LandCoverCode.WTLD
            Case Else : Return Nothing
        End Select
    End Function

End Class

File: GRMCore/Class/cSetLandcover.cs
using System;
using System.Data;

namespace GRMCore
{
    public class cSetLandcover
    {
        public enum LandCoverCode
        {
            WATR,
            URBN,
            BARE,
            WTLD,
            GRSS,
            FRST,
            AGRL,
            USER,
            CONSTV,
            NULL
        }

        public Nullable<cGRM.FileOrConst> mLandCoverDataType;
        public string mGridLandCoverFPN;
        public string mLandCoverVATFPN;
        public Dataset.GRMProject.LandCoverDataTable mdtLandCoverInfo;
        public Nullable<double> mConstRoughnessCoefficient;
        public Nullable<double> mConstImperviousRatio;

        public cSetLandcover()
        {
        }

        public void SetValues(Dataset.GRMProject prjDB)
        {
            if (mLandCoverDataType.HasValue)
            {
                Dataset.GRMProject.ProjectSettingsRow row = (Dataset.GRMProject.ProjectSettingsRow)prjDB.ProjectSettings.Rows[0];
                row.LandCoverDataType = mLandCoverDataType.Value.ToString();
                switch (mLandCoverDataType.Value)
                {
                    case cGRM.FileOrConst.File:
                        {
                            row.LandCoverFile = mGridLandCoverFPN;
                            row.LandCoverVATFile = mLandCoverVATFPN;
                            row.SetConstantRoughnessCoeffNull();
                            row.SetConstantImperviousRatioNull();
                            break;
                        }

                    case cGRM.FileOrConst.Constant:
                        {
                            row.SetLandCoverFileNull();
                            row.SetLandCoverVATFileNull();
                            row.ConstantRoughnessCoeff = System.Convert.ToString(mConstRoughnessCoefficient.Value);
                            row.ConstantImperviousRatio = System.Convert.ToString(mConstImperviousRatio.Value);
                            break;
                        }
                }
                if (mLandCoverDataType.Equals(cGRM.FileOrConst.File))
                {
                    mdtLandCoverInfo.AcceptChanges();
                    foreach (DataRow r in mdtLandCoverInfo.Rows)
                    {
                        r.SetAdded();
                    }
                }
            }
        }

        public void GetValues(Dataset.GRMProject prjDB)
        {
            mLandCoverDataType = default(cGRM.FileOrConst?);
            mLandCoverVATFPN = null;
            mConstRoughnessCoefficient = default(Single?);
            mConstImperviousRatio = default(Single?);
            Dataset.GRMProject.ProjectSettingsRow row = (Dataset.GRMProject.ProjectSettingsRow)prjDB.ProjectSettings.Rows[0];
            if (!row.IsLandCoverDataTypeNull())
            {
                if (row.LandCoverDataType.ToLower() == cGRM.FileOrConst.File.ToString().ToLower())
                {
                    mLandCoverDataType = cGRM.FileOrConst.File;
                    mGridLandCoverFPN = row.LandCoverFile;
                    mLandCoverVATFPN = row.LandCoverVATFile;
                }
                else if (row.LandCoverDataType.ToLower() == cGRM.FileOrConst.Constant.ToString().ToLower())
                {
                    mLandCoverDataType = cGRM.FileOrConst.Constant;
                    double v = 0;
                    if (double.TryParse(row.ConstantImperviousRatio, out v)) { mConstImperviousRatio = v; }
                    else if (double.TryParse(row.ConstantRoughnessCoeff, out v)) { mConstRoughnessCoefficient = v; }
                }
                else
                {
                    throw new InvalidOperationException();
                }
            }

            if (mLandCoverDataType.Equals(cGRM.FileOrConst.File))
            {
                mdtLandCoverInfo = new Dataset.GRMProject.LandCoverDataTable();
                mdtLandCoverInfo = prjDB.LandCover;
            }
            else
            { mdtLandCoverInfo = null; }
        }

        public bool IsSet
        {
            get
            {
                return mLandCoverDataType.HasValue;
            }
        }

        public string GetLandCoverName(int intGridValue)
        {
            if (mLandCoverDataType.Equals(cGRM.FileOrConst.File))
            {
                DataRow[] rows = mdtLandCoverInfo.Select("GridValue = " + intGridValue);
                return rows[0]["GRMLandCoverE"].ToString();
            }
            else
            {
                return "[CONST]";
            }
        }

        public static LandCoverCode GetLandCoverCode(string inName)
        {
            switch (inName.Trim())
            {
                case nameof(LandCoverCode.AGRL):
                    {
                        return cSetLandcover.LandCoverCode.AGRL;
                    }

                case nameof(LandCoverCode.BARE):
                    {
                        return cSetLandcover.LandCoverCode.BARE;
                    }

                case nameof(LandCoverCode.FRST):
                    {
                        return LandCoverCode.FRST;
                    }

                case nameof(LandCoverCode.GRSS):
                    {
                        return LandCoverCode.GRSS;
                    }

                case nameof(LandCoverCode.URBN):
                    {
                        return LandCoverCode.URBN;
                    }

                case nameof(LandCoverCode.WATR):
                    {
                        return LandCoverCode.WATR;
                    }

                case nameof(LandCoverCode.WTLD):
                    {
                        return LandCoverCode.WTLD;
                    }

                case nameof(LandCoverCode.USER):
                    {
                        return LandCoverCode.USER;
                    }

                case nameof(LandCoverCode.CONSTV):
                    {
                        return LandCoverCode.CONSTV;
                    }

                default:
                    {
                        return default(LandCoverCode);
                    }
            }
        }
    }
}
